### ISSUE TITLE
helloworld_cuda throws cuda driver insufficient error

### ISSUE THREAD

**Issue by feliiiiicity** [arrayfire/arrayfire#2410]
Hi, I'm using ubuntu 16.04 with cuda 9.2.148 and nvidia driver 396.44. When I ran the generated helloworld_cuda file, it returns:

ArrayFire Exception (Internal error:998):
In function cuda::DeviceManager::DeviceManager()
In file src/backend/cuda/platform.cpp:509
CUDA Error (35): CUDA driver version is insufficient for CUDA runtime version

In function void af::setDevice(int)
In file src/api/cpp/device.cpp:92
terminated called after throwing an instance of 'af::exception'
    what(): ArrayFire Exception(Internal error:998):
In function cuda::DeviceManager::DeviceManager()
In file src/backend/cuda/platform.cpp:509
CUDA Error (35): CUDA driver version is insufficient for CUDA runtime version

In function void af::setDevice(int)
In file src/api/cpp/device.cpp:92
Aborted (core dumped)

Do you have any suggestions? I installed cuda 9.2 and nvidia driver through runfiles.

**Comment by 9prady9** [comment]
Is ArrayFire installed using binary? Or built from source?

**Comment by pavanky** [comment]
@feliiiiicity try restarting your system.

**Comment by feliiiiicity** [comment]
> Is ArrayFire installed using binary? Or built from source?

I installed ArrayFire from the binary installer. I'ver tried restarting the system, but the problem is still there. I think the version of CUDA driver and runtime are matched, since the output of ./devuceQuery for CUDA prints "... CUDA Driver = CUDART, CUDA Driver Version =9.2, CUDA Runtime Version = 9.2, NumDevs = 2, Result = PASS".

**Comment by feliiiiicity** [comment]
@9prady9 @pavanky I installed ArrayFire from the binary installer. I'ver tried restarting the system, but the problem is still there. I think the version of CUDA driver and runtime are matched, since the output of ./devuceQuery for CUDA prints "... CUDA Driver = CUDART, CUDA Driver Version =9.2, CUDA Runtime Version = 9.2, NumDevs = 2, Result = PASS".

**Comment by 9prady9** [comment]
Current binary release is built with CUDA  10 so it requires 400 series driver

**Comment by feliiiiicity** [comment]
@9prady9 so, is there any binary installer suited for my system? or any other options ?  Thanks.

**Comment by 9prady9** [comment]
You should be able to use 3.6.1 binary release. However, if you want to have fixes from 3.6.2 , you can build from source if that is something you are interested in.

**Comment by 9prady9** [comment]
In fact, you are going to build from source, you can build from master to get all the most recent fixes.

**Comment by feliiiiicity** [comment]
@9prady9 Ok, thanks. I'll try building from source then.

**Comment by umar456** [comment]
ArrayFire provides its own CUDA runtime with the binary installers. You don't need a CUDA toolkit installed on your system to use ArrayFire. You only need the GPU drivers. You can find the drivers [here](https://www.nvidia.com/Download/index.aspx?lang=en-us) or you could add the graphics driver ppa and use the package manager to install the drivers.
```
add-apt-repository ppa:graphics-drivers/ppa
apt-get update
apt-get install nvidia-410 nvidia-settings
```

Is it not possible for you to update the drivers on your system? I am curious because I want to know what the barriers people are experiencing when using ArrayFire.

**Comment by feliiiiicity** [comment]
@umar456 Thanks a lot, I'll give you update on this. I need to talk to the supervisor first before updating the driver, since there're other users working on the server.

**Comment by feliiiiicity** [comment]
@umar456 @9prady9 @pavanky  I updated nvidia-driver to 410 and have passed the helloworld_cuda test. Thanks.

**Comment by WilliamTambellini** [comment]
Hi guys, I m still wondering if afcuda could check that the driver version
is adequate and if not correctly log/warn the desired minimum version.
I would be ok to work on it and pullrequest if we agree on some specs.
W.

**Comment by umar456** [comment]
I think a more specific message would be a good idea. @WilliamTambellini If you want to take this on, you can implement it this way:

You will probably modify the DeviceManager constructor here: https://github.com/arrayfire/arrayfire/blob/master/src/backend/cuda/platform.cpp#L472. I think this will be the first method called in any ArrayFire program. You can get the version of the driver by calling [`nvDriverVersion`](https://github.com/arrayfire/arrayfire/blob/master/src/backend/cuda/driver.cpp) or [nvmlSystemGetDriverVersion](https://docs.nvidia.com/deploy/nvml-api/group__nvmlSystemQueries.html#group__nvmlSystemQueries_1gee65c28b1a5c84408d61069a0097ba18)(better). This value can be used to index a table with known minimum driver versions for the different CUDA toolkit versions we support. I believe this is 7.5 at the moment. This table should be populated by the values in Table 1 on this page:

Let me know if you have additional questions.
https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html

If the versions do not match you should throw a AF_ERR_DRIVER error with a helpful message about what the user needs to do to download the latest versions.

**Comment by WilliamTambellini** [comment]
ok @umar456  I will take that one.

**Comment by WilliamTambellini** [comment]
ok so nvml does nt seem to be available in cuda 7 and is more over not searched/found by cmake (at least 3.9) : looks like cmake has still not yet moved forward on it :( 
https://gitlab.kitware.com/cmake/cmake/issues/17175
Trying another way ...

**Comment by WilliamTambellini** [comment]
https://github.com/arrayfire/arrayfire/pull/2421

**Comment by WilliamTambellini** [comment]
Hello @umar456 
I ve just tried the official 3.6.3 (built with/embedding cuda 10.0) available on
https://arrayfire.com/download/
https://arrayfire.s3.amazonaws.com/3.6.3/ArrayFire-v3.6.3_Linux_x86_64.sh
and the behavior is still not the expected one as I guess we agreed :  
- current behavior : AF unified failed to use afcuda so default to afopencl 
- expected behavior : AF to throw an exception that the nvidia drivers are too old for cuda (implemented in #2421 )

Example with AF_TRACE=all : 
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: 
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:139 ] Found: libafcpu.so.3
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:146 ] Device Count: 1.
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: 
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:139 ] Found: libafopencl.so.3
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:146 ] Device Count: 1.
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: 
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:139 ] Found: libafcuda.so.3
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:146 ] Device Count: 0.
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: .
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: ./src/backend/cuda
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: ../src/backend/cuda
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: src/backend/cuda
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: lib
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: lib64
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: 
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:139 ] Found: libafcuda.so.3
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:146 ] Device Count: 0.
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: /opt/arrayfire-3/lib/
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: /opt/arrayfire/lib/
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: /usr/local/lib/
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:137 ] Attempting: /usr/local/arrayfire/lib/
[unified][003987] [ ../src/api/unified/symbol_manager.cpp:194 ] AF_DEFAULT_BACKEND: opencl
ArrayFire v3.6.3 (OpenCL, 64-bit Linux, build 25bb360)
[0] INTEL: Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz, 15925 MB

$ nvidia-smi 
Tue Apr 23 09:59:37 2019       
| NVIDIA-SMI 396.82                 Driver Version: 396.82 

Before anyone ask : yes, I know that these drivers are too old for cuda 10 and yes I could upgrade, but that test is on purposely testing the capacity for AF to detect/throw to the user that the nvidia drivers are too old.

Could we at least reopen that issue ?

**Comment by umar456** [comment]
I am not sure how to proceed on this issue. If the other backends are working and CUDA is not then we should select the other backends. Throwing an exception will cause exceptions to be thrown on non-NVIDIA systems because the driver will not be available. By default this is the correct behavior.

If you select a particular backend, then that call should throw an error with a descriptive error message. We should also update the logs to show why a particular backend wasn't selected.
